### PR TITLE
Fix types definition issues

### LINKED ISSUE #377: Typings for new features (Event Engine & Listeners)
Hello everyone. 

Could someone please help to update typings:
https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/pubnub

Latest version is `"@types/pubnub": "7.4.2"` and PubNub `"pubnub": "8.0.1"`.

Especially need to have new methods from new **Event Engine & Listeners**, such as `subscriptionSet`. 

Thanks.

### LINKED ISSUE #405: Importing types from pubnub
starting 8.2.8 a major change was made (should not hide behind a patch version) that types are now used from internal library instead of using definitely typed. 

Can you please document on the proper way to import from the library? Also - are all the types that existed in DT still supported? 
It seems like some types are not exported.

### LINKED ISSUE #407: Incorrect history Actions type
The current typing for the history `Actions` interface is wrong. It defines the action value as an object when it is actually an array of objects.

https://github.com/pubnub/javascript/blob/0beda2cb0b80eba69309111fb27a76d8c6319081/src/core/types/api/history.ts#L134-L144

### LINKED ISSUE #412: PublishParameters message type too strict
It looks like the `Payload` type used by `PublishParameters.message` attempts to ensure the value contains only valid JSON types, but using this type on inputs prevents users from utilizing the JSON serialization layer. Namely, classes/objects with custom JSON serialization via `toJSON()`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior

For example, instances of `Date`. Another example is that `undefined` is not allowed even though JSON serialization excludes those properties automatically.

I suggest using `unknown` instead of `Payload`.

The workaround is to use the `any` trapdoor which defeats the purpose of the typing entirely.

### LINKED ISSUE #413: Parameter bag types with index signatures are too lax
In several places, inputs that would otherwise be strictly structured contain index signatures which allow the user to specify properties with any key. This prevents the typings from being able to catch misspellings and other mistakes. Specifically, in argument "bags".

For example, `PublishParameters` contains an index signature "for deprecated parameters", but this allows the following misspelling without any warning or error:
```typescript
pubnub.publish({channel: 'fu', message: 'bar', storeInHistor: false})
```

Without the index signature, Typescript would flag `storeInHistor` with the error: `Object literal may only specify known properties, but 'storeInHistor' does not exist in type 'PublishParameters'. Did you mean to write 'storeInHistory'?`

Rather than an index signature, deprecated params should be explicitly listed but marked as deprecated. Or, a dedicated subkey or alternate param should be provided for specifying arbitrary keys which are merged when sent to the API.

### PR DESCRIPTION
**Pull request by parfeon**
fix(types): fix `Actions` type definition
    
Fix definition of type which represents message actions received from history and list of users which added action of specific type and value to the message.
    
Closes #407
    
refactor(types): remove indexed signature for publish
    
Remove redundant indexed signature from publish message parameters type definition.
    
Closes #413
    
refactor(types): add serializable objects to `Payload` type
    
Extend `Payload` type definition with objects which can be serialized by `JSON.stringify` using `toJSON()` methods.
    
Closes #412
    
refactor(types): aggregate generated types definitions
    
Aggregate multiple types definitions into single type definition type with proper type names and namespaces.
    
Closes #405 #409 #410
    
refactor(types): add missing Subscribe Event Engine types
    
Add the Subscribe Event Engine and Event Listener types to the bundled types definition file.
    
Closes #377

### CODE REVIEW

File: .scripts/types-aggregate.ts [hunk 1]
```diff
@@ -0,0 +1,803 @@
+import { program } from 'commander';
```
Comment (mohitpubnub) [inline .scripts/types-aggregate.ts:1]: 👍🏻

File: src/core/types/api/index.ts [hunk 2]
```diff
@@ -142,4 +146,10 @@ export type Query = Record<string, string | number | (string | number)[]>;
  * * generic messages and signals content,
  * * published message metadata.
  */
-export type Payload = string | number | boolean | { [key: string]: Payload | null } | Payload[];
+export type Payload =
```
Comment (mohitpubnub) [inline src/core/types/api/index.ts:149]: Just re verifying. Does this Payload includes Date related types?

**Review by mohitpubnub — APPROVED** [review]

File: src/core/types/api/index.ts (hunk 2 above)
Comment (parfeon) [reply src/core/types/api/index.ts:149]: `Date` implements `.toJSON()` function, so it will work as well.

**Comment by parfeon** [comment]
@pubnub-release-bot release

**Pull request merged**